File: tests/CodexiumMagnus.Core.Tests/Configuration/CompositeConfigurationResolverTests.cs
using System.Collections.Generic;
using CodexiumMagnus.Core.Configuration;
using CodexiumMagnus.Core.Models;
using Xunit;

namespace CodexiumMagnus.Core.Tests.Configuration
{
    public class CompositeConfigurationResolverTests
    {
        // helper sources to make the tests readable
        private sealed class SimpleSource : IConfigurationSource
        {
            private readonly TypographyConfig? _typ;
            private readonly BibliographyConfig? _bib;

            public SimpleSource(TypographyConfig? typ = null, BibliographyConfig? bib = null)
            {
                _typ = typ;
                _bib = bib;
            }

            public TypographyConfig? GetTypography() => _typ;
            public BibliographyConfig? GetBibliography() => _bib;
        }

        [Fact]
        public void Typography_HigherLayer_OverridesLowerLayer()
        {
            // arrange
            var system = new SimpleSource(
                new TypographyConfig
                {
                    BaseFontFamily = "SystemFont",
                    BaseFontSizePt = 12
                });

            var user = new SimpleSource(
                new TypographyConfig
                {
                    BaseFontSizePt = 14 // user wants bigger
                });

            var session = new SimpleSource(
                new TypographyConfig
                {
                    BaseFontFamily = "SessionFont" // temporary override
                });

            var resolver = new CompositeConfigurationResolver(new List<IConfigurationSource>
            {
                // order: Session → User → Corpus → System
                session,
                user,
                system
            });

            // act
            var effective = resolver.GetEffectiveTypography();

            // assert
            Assert.Equal("SessionFont", effective.BaseFontFamily);
            Assert.Equal(14, effective.BaseFontSizePt);
        }

        [Fact]
        public void Bibliography_UsesSystem_WhenNoOtherLayers()
        {
            var system = new SimpleSource(
                bib: new BibliographyConfig
                {
                    Style = "APA",
                    SortBy = "author"
                });

            var resolver = new CompositeConfigurationResolver(new List<IConfigurationSource>
            {
                system
            });

            var effective = resolver.GetEffectiveBibliography();

            Assert.Equal("APA", effective.Style);
            Assert.Equal("author", effective.SortBy);
            Assert.Null(effective.GroupBy);
        }

        [Fact]
        public void Bibliography_Session_CanOverride_SortOnly()
        {
            var system = new SimpleSource(
                bib: new BibliographyConfig
                {
                    Style = "APA",
                    SortBy = "author"
                });

            var user = new SimpleSource(
                bib: new BibliographyConfig
                {
                    Style = "CMS" // user prefers CMS
                });

            var session = new SimpleSource(
                bib: new BibliographyConfig
                {
                    SortBy = "year" // temporary view
                });

            var resolver = new CompositeConfigurationResolver(new List<IConfigurationSource>
            {
                session,
                user,
                system
            });

            var effective = resolver.GetEffectiveBibliography();

            // session wins for sort
            Assert.Equal("year", effective.SortBy);
            // user wins for style
            Assert.Equal("CMS", effective.Style);
        }
    }
}

File: tests/codexium-magnus-core-tests/Configuration/CompositeConfigurationResolverTests.cpp
#include <QtTest/QtTest>
#include "Configuration/CompositeConfigurationResolver.h"
#include "Configuration/ConfigurationSource.h"
#include "Models/TypographyConfig.h"
#include "Models/BibliographyConfig.h"

using namespace CodexiumMagnus::Core::Configuration;
using namespace CodexiumMagnus::Core::Models;

// Helper source for testing
class SimpleSource : public ConfigurationSource {
public:
    SimpleSource(TypographyConfig* typ = nullptr, BibliographyConfig* bib = nullptr)
        : m_typ(typ), m_bib(bib) {}

    TypographyConfig* getTypography() override { return m_typ; }
    BibliographyConfig* getBibliography() override { return m_bib; }

private:
    TypographyConfig* m_typ;
    BibliographyConfig* m_bib;
};

class CompositeConfigurationResolverTests : public QObject {
    Q_OBJECT

private slots:
    void typography_HigherLayer_OverridesLowerLayer();
    void bibliography_UsesSystem_WhenNoOtherLayers();
    void bibliography_Session_CanOverride_SortOnly();
};

void CompositeConfigurationResolverTests::typography_HigherLayer_OverridesLowerLayer() {
    // Arrange
    auto systemTyp = new TypographyConfig();
    systemTyp->baseFontFamily = "SystemFont";
    systemTyp->baseFontSizePt = 12;
    SimpleSource system(systemTyp, nullptr);

    auto userTyp = new TypographyConfig();
    userTyp->baseFontSizePt = 14; // user wants bigger
    SimpleSource user(userTyp, nullptr);

    auto sessionTyp = new TypographyConfig();
    sessionTyp->baseFontFamily = "SessionFont"; // temporary override
    SimpleSource session(sessionTyp, nullptr);

    QList<ConfigurationSource*> sources = {&session, &user, &system};
    CompositeConfigurationResolver resolver(sources);

    // Act
    TypographyConfig effective = resolver.getEffectiveTypography();

    // Assert
    QCOMPARE(effective.baseFontFamily, QString("SessionFont"));
    QCOMPARE(effective.baseFontSizePt, 14.0);
}

void CompositeConfigurationResolverTests::bibliography_UsesSystem_WhenNoOtherLayers() {
    auto systemBib = new BibliographyConfig();
    systemBib->style = "APA";
    systemBib->sortBy = "author";
    SimpleSource system(nullptr, systemBib);

    QList<ConfigurationSource*> sources = {&system};
    CompositeConfigurationResolver resolver(sources);

    BibliographyConfig effective = resolver.getEffectiveBibliography();

    QCOMPARE(effective.style, QString("APA"));
    QCOMPARE(effective.sortBy, QString("author"));
    QVERIFY(effective.groupBy.isEmpty());
}

void CompositeConfigurationResolverTests::bibliography_Session_CanOverride_SortOnly() {
    auto systemBib = new BibliographyConfig();
    systemBib->style = "APA";
    systemBib->sortBy = "author";
    SimpleSource system(nullptr, systemBib);

    auto userBib = new BibliographyConfig();
    userBib->style = "CMS"; // user prefers CMS
    SimpleSource user(nullptr, userBib);

    auto sessionBib = new BibliographyConfig();
    sessionBib->sortBy = "year"; // temporary view
    SimpleSource session(nullptr, sessionBib);

    QList<ConfigurationSource*> sources = {&session, &user, &system};
    CompositeConfigurationResolver resolver(sources);

    BibliographyConfig effective = resolver.getEffectiveBibliography();

    // Session wins for sort
    QCOMPARE(effective.sortBy, QString("year"));
    // User wins for style
    QCOMPARE(effective.style, QString("CMS"));
}

QTEST_MAIN(CompositeConfigurationResolverTests)
#include "CompositeConfigurationResolverTests.moc"
